src/lib/models/containers.ts

```
@@ -0,0 +1,485 @@
+	private containerIdToTraceId = new Map<string, string>(); // containerId -> traceId
+
+	addContainer(container: Container): void {
+		this.allContainers.set(container.id, container);
+
+		const traceId = container.getTraceId();
+		if (traceId) {
+			// Track which containers belong to which trace
+			if (!this.traceIdToContainerIds.has(traceId)) {
+				this.traceIdToContainerIds.set(traceId, new Set());
+			}
+			this.traceIdToContainerIds.get(traceId)!.add(container.id);
```

Biome (1.9.4)

[error] Forbidden non-null assertion.

Unsafe fix: Replace with optional chain operator ?. This operator includes runtime checks, so it is safer than the compile-only non-null assertion operator


(lint/style/noNonNullAssertion)